src/proof/proof_rule.h

```diff
@@ -1385,241 +1385,349 @@ enum class PfRule : uint32_t
+   * Also note that constants are split, such that if ``P`` concludes :math:`(x
+   * \cdot \mathsf{'abc'}) = (y \cdot \mathsf{'a'})`, then ``(CONCAT_EQ P :args
+   * false)`` concludes :math:`(x \cdot \mathsf{'bc'}) = y`.  This splitting is
+   * done only for constants such that ``Word::splitConstant`` returns non-null.
+   * \endverbatim
+   */
   CONCAT_EQ,
-  // ======== Concat unify
-  // Children: (P1:(= (str.++ t1 t2) (str.++ s1 s2)),
-  //            P2:(= (str.len t1) (str.len s1)))
-  // Arguments: (b), indicating if reverse direction
-  // ---------------------
-  // Conclusion: (= t1 s1)
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Stings -- Core rules -- Concatenation unification**
+   *
+   * .. math::
+   *
+   *   \inferrule{(t_1\cdot t_2) = (s_1 \cdot s_2),\, \mathit{len}(t_1) =
+   *   \mathit{len}(s_1)\mid b}{t_1 = s_1}
+   *
+   * where :math:`b` indicates if the direction is reversed.
+   * \endverbatim
+   */
   CONCAT_UNIFY,
-  // ======== Concat conflict
-  // Children: (P1:(= (str.++ c1 t) (str.++ c2 s)))
-  // Arguments: (b), indicating if reverse direction
-  // ---------------------
-  // Conclusion: false
-  // Where c1, c2 are constants such that Word::splitConstant(c1,c2,index,b)
-  // is null, in other words, neither is a prefix of the other.
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Stings -- Core rules -- Concatenation conflict**
+   *
+   * .. math::
+   *
+   *   \inferrule{(c_1\cdot t) = (c_2 \cdot s)\mid b}{\bot}
+   *
+   * where :math:`b` indicates if the direction is reversed, :math:`c_1,\,c_2`
+   * are constants such that :math:`\texttt{Word::splitConstant}(c_1,c_2,
+   * \mathit{index},b)` is null, in other words, neither is a prefix of the
+   * other.
+   * \endverbatim
+   */
   CONCAT_CONFLICT,
-  // ======== Concat split
-  // Children: (P1:(= (str.++ t1 t2) (str.++ s1 s2)),
-  //            P2:(not (= (str.len t1) (str.len s1))))
-  // Arguments: (false)
-  // ---------------------
-  // Conclusion: (or (= t1 (str.++ s1 r_t)) (= s1 (str.++ t1 r_s)))
-  // where
-  //   r_t = (skolem (suf t1 (str.len s1)))),
-  //   r_s = (skolem (suf s1 (str.len t1)))).
-  //
-  // or the reverse form of the above:
-  //
-  // Children: (P1:(= (str.++ t1 t2) (str.++ s1 s2)),
-  //            P2:(not (= (str.len t2) (str.len s2))))
-  // Arguments: (true)
-  // ---------------------
-  // Conclusion: (or (= t2 (str.++ r_t s2)) (= s2 (str.++ r_s t2)))
-  // where
-  //   r_t = (skolem (pre t2 (- (str.len t2) (str.len s2))))),
-  //   r_s = (skolem (pre s2 (- (str.len s2) (str.len t2))))).
-  //
-  // Above, (suf x n) is shorthand for (str.substr x n (- (str.len x) n)) and
-  // (pre x n) is shorthand for (str.substr x 0 n).
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Stings -- Core rules -- Concatenation split**
+   *
+   * .. math::
+   *
+   *   \inferrule{(t_1\cdot t_2) = (s_1 \cdot s_2),\,
+   *   \mathit{len}(t_1) \neq \mathit{len}(s_1)\mid \bot}{(t_1 = s_1\cdot r_t)
```

Comment:
Do we actually require a `false` argument for this rule?

Reply:
Agreed.

Reply:
I'm only doing this for `CONCAT_SPLIT` because only the argument changes. There are a few other rules where we have this pattern of arg being true or false but there the premises/conclusions change, where it seems strange to separate just via side condition. It actually seems strange for them to be the same rule...